lms/envs/test.py

```diff
@@ -636,3 +636,17 @@
 #################### Network configuration ####################
 # Tests are not behind any proxies
 CLOSEST_CLIENT_IP_FROM_HEADERS = []
+
+COURSE_ENROLLMENT_MODES['test'] = {
+    "id": 8,
+    "slug": "test",
+    "display_name": "Test",
+    "min_price": 0
+}
+
+COURSE_ENROLLMENT_MODES['test_mode'] = {
+    "id": 9,
+    "slug": "test_mode",
+    "display_name": "Test Mode",
+    "min_price": 0
+}
```

Comment:
This commit (83811298d5b8ed65d6fc50a0ea5bb6acb9993d6c) does not seem to be using it anywhere.

Reply:
@Agrendalath, I'm not sure why email tests aren't failing for the upstream, but looks like we need this, after all.

<details>
  <summary>Failed tests output.</summary>

  ```python
=================================== FAILURES ===================================
________ TestEmailSendFromDashboardMockedHtmlToText.test_send_to_track _________

self = <lms.djangoapps.bulk_email.tests.test_email.TestEmailSendFromDashboardMockedHtmlToText testMethod=test_send_to_track>

    def test_send_to_track(self):
        """
        Make sure email sent to a registration track goes there.
        """
        CourseMode.objects.create(mode_slug='test', course_id=self.course.id)
        for student in self.students:
>           update_enrollment(student, str(self.course.id), 'test')

lms/djangoapps/bulk_email/tests/test_email.py:387: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
openedx/core/djangoapps/enrollments/api.py:321: in update_enrollment
    enrollment = _data_api().update_course_enrollment(username, course_id, mode=mode, is_active=is_active)
openedx/core/djangoapps/enrollments/data.py:189: in update_course_enrollment
    return _update_enrollment(enrollment, is_active=is_active, mode=mode)
openedx/core/djangoapps/enrollments/data.py:282: in _update_enrollment
    return CourseEnrollmentSerializer(enrollment).data
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/serializers.py:548: in data
    ret = super().data
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/serializers.py:246: in data
    self._data = self.to_representation(self.instance)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/serializers.py:515: in to_representation
    ret[field.field_name] = field.to_representation(attribute)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/fields.py:1882: in to_representation
    return method(value)
openedx/core/djangoapps/enrollments/serializers.py:118: in get_grading
    for section in course_grade.summary.get('section_breakdown'):
lms/djangoapps/grades/course_grade.py:189: in summary
    grade_summary = self.grader_result()
lms/djangoapps/grades/course_grade.py:177: in grader_result
    self.graded_subsections_by_format(visible_grades_only=visible_grades_only,
lms/djangoapps/grades/course_grade.py:92: in graded_subsections_by_format
    for chapter in self.chapter_grades.values():
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/lazy/lazy.py:28: in __get__
    value = self.__func(inst)
lms/djangoapps/grades/course_grade.py:109: in chapter_grades
    course_structure = self.course_data.structure
lms/djangoapps/grades/course_data.py:59: in structure
    self._structure = get_course_blocks(
lms/djangoapps/course_blocks/api.py:107: in get_course_blocks
    return get_block_structure_manager(starting_block_usage_key.course_key).get_transformed(
openedx/core/djangoapps/content/block_structure/manager.py:78: in get_transformed
    transformers.transform(block_structure)
openedx/core/djangoapps/content/block_structure/transformers.py:114: in transform
    self._transform_without_filters(block_structure)
openedx/core/djangoapps/content/block_structure/transformers.py:140: in _transform_without_filters
    transformer.transform(self.usage_info, block_structure)
lms/djangoapps/course_blocks/transformers/user_partitions.py:91: in transform
    user_groups = get_user_partition_groups(usage_info.course_key, user_partitions, user, 'id')
common/lib/xmodule/xmodule/partitions/partitions_service.py:65: in get_user_partition_groups
    group = partition.scheme.get_group_for_user(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'openedx.core.djangoapps.verified_track_content.partition_scheme.EnrollmentTrackPartitionScheme'>
course_key = CourseLocator('org.125', 'course_125', 'ẗëṡẗ_title_ｲ乇丂ｲ_ﾶ乇丂丂ﾑg乇_ｷo尺_ﾑﾚﾚ_тэѕт_мэѕѕаБэ', None, None)
user = <User: robot597>
user_partition = EnrollmentTrackUserPartition(id=50, name='Enrollment Track Groups', description='Partition for segmenting users by enr...'>, parameters={'course_id': 'course-v1:org.125+course_125+ẗëṡẗ_title_ｲ乇丂ｲ_ﾶ乇丂丂ﾑg乇_ｷo尺_ﾑﾚﾚ_тэѕт_мэѕѕаБэ'}, active=True)
kwargs = {}, mode_slug = 'test', is_active = True
course_mode = Mode(slug='test', name='', min_price=0, suggested_prices='', currency='usd', expiration_datetime=None, description=None, sku=None, bulk_sku=None)

    @classmethod
    def get_group_for_user(cls, course_key, user, user_partition, **kwargs):  # pylint: disable=unused-argument
        """
        Returns the Group from the specified user partition to which the user
        is assigned, via enrollment mode. If a user is in a Credit mode, the Verified or
        Professional mode for the course is returned instead.
    
        If a course is using the Verified Track Cohorting pilot feature, this method
        returns None regardless of the user's enrollment mode.
        """
        if is_course_using_cohort_instead(course_key):
            return None
    
        # First, check if we have to deal with masquerading.
        # If the current user is masquerading as a specific student, use the
        # same logic as normal to return that student's group. If the current
        # user is masquerading as a generic student in a specific group, then
        # return that group.
        if get_course_masquerade(user, course_key) and not is_masquerading_as_specific_student(user, course_key):
            return get_masquerading_user_group(course_key, user, user_partition)
    
        mode_slug, is_active = CourseEnrollment.enrollment_mode_for_user(user, course_key)
        if mode_slug and is_active:
            course_mode = CourseMode.mode_for_course(
                course_key,
                mode_slug,
                modes=CourseMode.modes_for_course(course_key, include_expired=True, only_selectable=False),
            )
            if course_mode and CourseMode.is_credit_mode(course_mode):
                # We want the verified track even if the upgrade deadline has passed, since we
                # are determining what content to show the user, not whether the user can enroll
                # in the verified track.
                course_mode = CourseMode.verified_mode_for_course(course_key, include_expired=True)
            if not course_mode:
                course_mode = CourseMode.DEFAULT_MODE
>           return Group(ENROLLMENT_GROUP_IDS[course_mode.slug]["id"], str(course_mode.name))
E           KeyError: 'test'

openedx/core/djangoapps/verified_track_content/partition_scheme.py:98: KeyError
----------------------------- Captured stdout call -----------------------------
Installed 2 object(s) from 1 fixture(s)
------------------------------ Captured log call -------------------------------
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.use_legacy_frontend' accessed without a request, which is likely in the context of a celery task.
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Flag 'courseware.mfe_progress_milestones' accessed without a request, which is likely in the context of a celery task.
ERROR    openedx.core.djangoapps.safe_sessions.middleware:middleware.py:179 SafeCookieData BWC parse error: 'd80dxbhqtlcxpu1kw721p8ckn56m70rx'.
WARNING  common.djangoapps.track.contexts:contexts.py:38 unable to parse course_id "course-v1:org.125+course_125+%E1%BA%97%C3%AB%E1%B9%A1%E1%BA%97_title_%EF%BD%B2%E4%B9%87%E4%B8%82%EF%BD%B2_%EF%BE%B6%E4%B9%87%E4%B8%82%E4%B8%82%EF%BE%91g%E4%B9%87_%EF%BD%B7o%E5%B0%BA_%EF%BE%91%EF%BE%9A%EF%BE%9A_%D1%82%D1%8D%D1%95%D1%82_%D0%BC%D1%8D%D1%95%D1"
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/opaque_keys/__init__.py", line 193, in from_string
    return key_class._from_string(rest)
  File "/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/opaque_keys/edx/locator.py", line 274, in _from_string
    parse = cls.parse_url(serialized)
  File "/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/opaque_keys/edx/locator.py", line 141, in parse_url
    raise InvalidKeyError(cls, string)
opaque_keys.InvalidKeyError: <class 'opaque_keys.edx.locator.CourseLocator'>: org.125+course_125+%E1%BA%97%C3%AB%E1%B9%A1%E1%BA%97_title_%EF%BD%B2%E4%B9%87%E4%B8%82%EF%BD%B2_%EF%BE%B6%E4%B9%87%E4%B8%82%E4%B8%82%EF%BE%91g%E4%B9%87_%EF%BD%B7o%E5%B0%BA_%EF%BE%91%EF%BE%9A%EF%BE%9A_%D1%82%D1%8D%D1%95%D1%82_%D0%BC%D1%8D%D1%95%D1%95%D0%B0%D0%91%D1%8D

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/edx-platform/edx-platform/common/djangoapps/track/contexts.py", line 36, in course_context_from_url
    course_id = CourseKey.from_string(course_id_string)
  File "/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/opaque_keys/__init__.py", line 196, in from_string
    return cls.deprecated_fallback._from_deprecated_string(serialized)
  File "/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/opaque_keys/edx/locator.py", line 389, in _from_deprecated_string
    raise InvalidKeyError(cls, serialized)
opaque_keys.InvalidKeyError: <class 'opaque_keys.edx.locator.CourseLocator'>: course-v1:org.125+course_125+%E1%BA%97%C3%AB%E1%B9%A1%E1%BA%97_title_%EF%BD%B2%E4%B9%87%E4%B8%82%EF%BD%B2_%EF%BE%B6%E4%B9%87%E4%B8%82%E4%B8%82%EF%BE%91g%E4%B9%87_%EF%BD%B7o%E5%B0%BA_%EF%BE%91%EF%BE%9A%EF%BE%9A_%D1%82%D1%8D%D1%95%D1%82_%D0%BC%D1%8D%D1%95%D1%95%D0%B0%D0%91%D1%8D
WARNING  edx_toggles.toggles.internal.waffle.flag:flag.py:79 Grades: Flag 'grades.enforce_freeze_grade_after_course_end' accessed without a request, which is likely in the context of a celery task.
_ TestEmailSendFromDashboardMockedHtmlToText.test_send_to_track_other_enrollments _

self = <lms.djangoapps.bulk_email.tests.test_email.TestEmailSendFromDashboardMockedHtmlToText testMethod=test_send_to_track_other_enrollments>

    def test_send_to_track_other_enrollments(self):
        """
        Failing test for EDUCATOR-217: verifies that emails are only sent to
        users in a specific track if they're in that track in the course the
        email is being sent from.
        """
        # Create a mode and designate an enrolled user to be placed in that mode
        CourseMode.objects.create(mode_slug='test_mode', course_id=self.course.id)
        test_mode_student = self.students[0]
>       update_enrollment(test_mode_student, str(self.course.id), 'test_mode')

lms/djangoapps/bulk_email/tests/test_email.py:408: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
openedx/core/djangoapps/enrollments/api.py:321: in update_enrollment
    enrollment = _data_api().update_course_enrollment(username, course_id, mode=mode, is_active=is_active)
openedx/core/djangoapps/enrollments/data.py:189: in update_course_enrollment
    return _update_enrollment(enrollment, is_active=is_active, mode=mode)
openedx/core/djangoapps/enrollments/data.py:282: in _update_enrollment
    return CourseEnrollmentSerializer(enrollment).data
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/serializers.py:548: in data
    ret = super().data
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/serializers.py:246: in data
    self._data = self.to_representation(self.instance)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/serializers.py:515: in to_representation
    ret[field.field_name] = field.to_representation(attribute)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/rest_framework/fields.py:1882: in to_representation
    return method(value)
openedx/core/djangoapps/enrollments/serializers.py:118: in get_grading
    for section in course_grade.summary.get('section_breakdown'):
lms/djangoapps/grades/course_grade.py:189: in summary
    grade_summary = self.grader_result()
lms/djangoapps/grades/course_grade.py:177: in grader_result
    self.graded_subsections_by_format(visible_grades_only=visible_grades_only,
lms/djangoapps/grades/course_grade.py:92: in graded_subsections_by_format
    for chapter in self.chapter_grades.values():
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/lazy/lazy.py:28: in __get__
    value = self.__func(inst)
lms/djangoapps/grades/course_grade.py:109: in chapter_grades
    course_structure = self.course_data.structure
lms/djangoapps/grades/course_data.py:59: in structure
    self._structure = get_course_blocks(
lms/djangoapps/course_blocks/api.py:107: in get_course_blocks
    return get_block_structure_manager(starting_block_usage_key.course_key).get_transformed(
openedx/core/djangoapps/content/block_structure/manager.py:78: in get_transformed
    transformers.transform(block_structure)
openedx/core/djangoapps/content/block_structure/transformers.py:114: in transform
    self._transform_without_filters(block_structure)
openedx/core/djangoapps/content/block_structure/transformers.py:140: in _transform_without_filters
    transformer.transform(self.usage_info, block_structure)
lms/djangoapps/course_blocks/transformers/user_partitions.py:91: in transform
    user_groups = get_user_partition_groups(usage_info.course_key, user_partitions, user, 'id')
common/lib/xmodule/xmodule/partitions/partitions_service.py:65: in get_user_partition_groups
    group = partition.scheme.get_group_for_user(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'openedx.core.djangoapps.verified_track_content.partition_scheme.EnrollmentTrackPartitionScheme'>
course_key = CourseLocator('org.125', 'course_125', 'ẗëṡẗ_title_ｲ乇丂ｲ_ﾶ乇丂丂ﾑg乇_ｷo尺_ﾑﾚﾚ_тэѕт_мэѕѕаБэ', None, None)
user = <User: robot611>
user_partition = EnrollmentTrackUserPartition(id=50, name='Enrollment Track Groups', description='Partition for segmenting users by enr...'>, parameters={'course_id': 'course-v1:org.125+course_125+ẗëṡẗ_title_ｲ乇丂ｲ_ﾶ乇丂丂ﾑg乇_ｷo尺_ﾑﾚﾚ_тэѕт_мэѕѕаБэ'}, active=True)
kwargs = {}, mode_slug = 'test_mode', is_active = True
course_mode = Mode(slug='test_mode', name='', min_price=0, suggested_prices='', currency='usd', expiration_datetime=None, description=None, sku=None, bulk_sku=None)

    @classmethod
    def get_group_for_user(cls, course_key, user, user_partition, **kwargs):  # pylint: disable=unused-argument
        """
        Returns the Group from the specified user partition to which the user
        is assigned, via enrollment mode. If a user is in a Credit mode, the Verified or
        Professional mode for the course is returned instead.
    
        If a course is using the Verified Track Cohorting pilot feature, this method
        returns None regardless of the user's enrollment mode.
        """
        if is_course_using_cohort_instead(course_key):
            return None
    
        # First, check if we have to deal with masquerading.
        # If the current user is masquerading as a specific student, use the
        # same logic as normal to return that student's group. If the current
        # user is masquerading as a generic student in a specific group, then
        # return that group.
        if get_course_masquerade(user, course_key) and not is_masquerading_as_specific_student(user, course_key):
            return get_masquerading_user_group(course_key, user, user_partition)
    
        mode_slug, is_active = CourseEnrollment.enrollment_mode_for_user(user, course_key)
        if mode_slug and is_active:
            course_mode = CourseMode.mode_for_course(
                course_key,
                mode_slug,
                modes=CourseMode.modes_for_course(course_key, include_expired=True, only_selectable=False),
            )
            if course_mode and CourseMode.is_credit_mode(course_mode):
                # We want the verified track even if the upgrade deadline has passed, since we
                # are determining what content to show the user, not whether the user can enroll
                # in the verified track.
                course_mode = CourseMode.verified_mode_for_course(course_key, include_expired=True)
            if not course_mode:
                course_mode = CourseMode.DEFAULT_MODE
>           return Group(ENROLLMENT_GROUP_IDS[course_mode.slug]["id"], str(course_mode.name))
E           KeyError: 'test_mode'

openedx/core/djangoapps/verified_track_content/partition_scheme.py:98: KeyError
  ```
</details>

Also, I noticed that IDs [are outdated](https://github.com/openedx/edx-platform/blob/1d618055dc5a44b94a1da63ef4017a2015aad018/lms/envs/common.py#L4618), and have to be bumped to `13` and `14`.

Reply:
@0x29a, got it, let's not spend more time on this, then 👍🏻